Source/Core/InputCommon/ControllerInterface/Pipes/Pipes.h

```diff
@@ -0,0 +1,67 @@
+// Copyright 2015 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <map>
+#include <string>
+#include <vector>
+
+#include "InputCommon/ControllerInterface/Device.h"
+
+namespace ciface
+{
+namespace Pipes
+{
+// To create a piped controller input, create a named pipe in the
+// Pipes directory and write commands out to it. Commands are separated
+// by a newline character, with spaces separating command tokens.
+// Command syntax is as follows, where curly brackets are one-of and square
+// brackets are inclusive numeric ranges. Cases are sensitive. Numeric inputs
+// are clamped to [0, 1] and otherwise invalid commands are discarded.
+// {PRESS, RELEASE} {A, B, X, Y, Z, START, L, R, D_UP, D_DOWN, D_LEFT, D_RIGHT}
+// SET {L, R} [0, 1]
+// SET {MAIN, C} [0, 1] [0, 1]
+
+void Init(std::vector<Core::Device*>& devices);
+
+class PipeDevice : public Core::Device
+{
+private:
+	class PipeInput : public Input
+	{
+	public:
+		std::string GetName() const { return m_name; }
+		PipeInput(std::string name) : m_name(name), m_state(0.0) {}
+		ControlState GetState() const { return m_state; }
+		void SetState(ControlState state) { m_state = state; }
+	private:
+		const std::string m_name;
+		ControlState m_state;
+	};
+
+	void AddAxis(std::string name, double value);
+	void ParseCommand(std::string command);
+	void SetAxis(std::string entry, double value);
+
+	const int m_fd;
+	std::string m_name;
+	const int m_id;
+	std::string m_buf;
+	std::map<std::string, PipeInput*> m_buttons;
+	std::map<std::string, PipeInput*> m_axes;
+
+public:
+	void UpdateInput() override;
+
+	PipeDevice(int fd, const std::string& name, int id);
+	~PipeDevice();
```

Comment:
Constructor and destructor should be the first thing in the public section of the class